### PR TITLE
Only update queue items asynchronously

### PR DESCRIPTION
**Pull request by fredrikekelund**
### PR Checklist

- [x] I have read and/or are familiar with the contributor guidelines
- [x] I have checked to make sure no existing PRs already satisfy the original requirements of this PR
- [x] The commit messages in this PR are clear, and any WIP commits have been squashed
- [x] The code passes lint checking
- [x] The PR contains no spelling or grammatical errors
- [x] Any relevant documentation has been updated
- [x] I have included tests for my code

### What this PR changes

The final puzzle piece that enables asynchronous queue
implementations! It adds an `update` method to the queue that
takes a URL, an object that describes the updates to be made to
the queue item, and a callback.

I've also made sure that all the places where the crawler mutates
queueItems that are *in the queue* now use this method.

And of course, I've added some tests.

### Rationale

See #257

### CODE REVIEW

File: lib/crawler.js [hunk 1]
```diff
@@ -1062,98 +1062,97 @@ Crawler.prototype.queueURL = function(url, referrer) {
...
+    var crawler = this,
+        timeCommenced = Date.now();
 
-    // Up the socket limit if required.
-    if (agent.maxSockets < crawler.maxConcurrency) {
-        agent.maxSockets = crawler.maxConcurrency;
-    }
+    crawler.queue.update(queueItem.url, {
+        status: "spooled"
+    }, function(error, queueItem) {
```
Comment (cgiffard) [inline lib/crawler.js]: How do we handle the error here? :-/

File: test/queue.js [hunk 2]
```diff
@@ -188,4 +209,85 @@ describe("Queue methods", function() {
...
+        crawler.maxDepth = 2;
+
+        function findByUrl(array, url) {
+            return find(array, function(queueItem) {
+                return queueItem.url === url;
+            });
+        }
+
+        crawler.queue.add = function (queueItem) {
+            var storedQueueItem = deepAssign({}, queueItem);
```
Comment (cgiffard) [inline test/queue.js]: I like this. Maybe we can add small `setImmediate` calls to verify that no race conditions are occurring (well, increase the likelihood that we'd catch them here.)

**Comment by cgiffard** [comment]
Like it so far! I'll finish reviewing tonight. Thanks! :)

File: lib/crawler.js (hunk 1 above)
Comment (fredrikekelund) [inline lib/crawler.js]: Ah, you're right - there are a few errors we need to handle. Just to clarify a few things about these errors:
I thought initially about retry strategies for the update method and if that was something we ought to have, but I realized that that's only needed for third party queue implementations, and they can handle retry strategies internally. 
So when we do get an error here, do we just throw it or do we emit some kind of event? We can't really go ahead with what we were doing in any case, and the risk of frequent errors from the update method should be very low (I don't even think the default method can return one), so maybe just throw it?

File: test/queue.js (hunk 2 above)
Comment (fredrikekelund) [inline test/queue.js]: Sure, we could wrap it in some setImmediate calls, but is there any particular scenario where we run the risk of a race condition here.?

Comment (cgiffard) [inline test/queue.js]: No specific reason, but it's worth simulating an asynchronous queue in our tests.

File: lib/crawler.js (hunk 1 above)
Comment (cgiffard) [inline lib/crawler.js]: I'm torn between throwing it and emitting it. I think generally we don't want anything to crash the crawler, as there's the potential for lost state. So perhaps emitting an error would be best? Maybe `queueerror` or something?

Comment (fredrikekelund) [inline lib/crawler.js]: True, throwing the error is a bit like copping out, I guess. What about the fetching process - do we emit an event and then simply abort the fetching? And leave the queue item as it is?

File: test/queue.js (hunk 2 above)
Comment (fredrikekelund) [inline test/queue.js]: Ah, that's true - let's wrap it in an `setImmediate` call!

File: lib/crawler.js (hunk 1 above)
Comment (cgiffard) [inline lib/crawler.js]: I guess if we emit an error, and leave the queue item as it is, at some point the crawler will attempt to pick it up again (and if it's in the process of downloading — that is, the status as held by the queue is something other than `spooled` — it's the client's responsibility to keep an eye on the queue event to make sure that it's being handled appropriately. There's only so much we can do when the error handling strategy for the queue could be, well, anything.)

Comment (fredrikekelund) [inline lib/crawler.js]: Alright, so we emit an `queueerror` event or something similar, abort the fetching and rely on the queue to pick up the queue item again in the future, at which point it's hopefully recovered and is able to complete the operation. Sounds like a good plan of action!

**Comment by fredrikekelund** [comment]
All done with the async tests and error handling (as well as a test for that error handling)! Hopefully we should be good to merge now, I'll let you have a final look @cgiffard

**Comment by brendonboshell** [comment]
I'm looking at the `Crawler#start` function ([crawler.js#L302](https://github.com/fredrikekelund/node-simplecrawler/blob/asyncQueueItemUpdate/lib/crawler.js#L302)), for instance. It appears to call the `add` method, but then calls `crawl` within the next tick. But, at that point in time, the queue could will be empty (because `add` could be running asynchronously), and it appears the crawl would stop.

Indeed, if I take the existing (synchronous logic) in `FetchQueue#add` and wrap it in a setTimeout, the crawl immediately stops before it even starts.

I haven't check if there are similar issues elsewhere in the codebase.

**Comment by cgiffard** [comment]
Good catch. There's definitely a race condition introduced by gzipping in `master` now, so it might be worth trying to kill two birds with one stone, and reviewing this for more. :-/

I think we need to do something equivalent to `crawler.wait` (or just generalise `crawler.wait` to support this case too) every time we fire up an async function, and resolve it when the response comes back (wait conditions time out, so they seem like a good base to start from.) Using that single method should hopefully prevent race conditions (well, race conditions involving async tasks which take less than the timeout threshold.)

**Comment by cgiffard** [comment]
For reference, the other race condition is described by #298.

**Comment by fredrikekelund** [comment]
@brendonboshell thanks for taking the time to review the changes here! I've updated the code to prevent the race condition (and made sure that the #298 problem remains fixed in this branch as well).

**Comment by fredrikekelund** [comment]
I'd like to try and merge this later tonight unless you have any objections, @cgiffard? I'll fix any merge conflicts in #293 and then go ahead with that one as well, I think

**Comment by cgiffard** [comment]
@fredrikekelund Thanks! This looks great. :)

**Comment by fredrikekelund** [comment]
I realized just after merging this that the crawler.update method shouldn't rely on a URL as the first argument. There can be multiple queue items with the same URL in the queue, so we should have an id property on all queue items after all. In the default queue implementation that could just be the index of the queue item in the queue (but we still need to store it on the queue item as well). This should be a fairly simple change all in all, though!

**Pull request merged**